ang/civiawards-workflow/services/applicant-management-workflow.service.js

```diff
@@ -0,0 +1,156 @@
+
+          return civicaseCrmApi('AwardDetail', 'get', filters)
+            .then(function (awardsData) {
+              return _.map(awardsData.values, function (award) {
+                return award.case_type_id;
+              });
+            });
+        });
+    }
+
+    /**
+     * Returns workflows list for applicant management
+     *
+     * @param {Array} scope scope object of the workflows list controller
+     * @returns {Array} api call parameters
+     */
+    function getWorkflowsList (scope) {
+      return applyFilter(scope)
+        .then(function (caseTypeIds) {
+          var params = {
+            sequential: 1,
+            is_active: scope.selectedFilters.is_active,
+            case_type_category: scope.caseTypeCategory,
+            options: { limit: 0 },
+            'api.AwardDetail.get': {
+              case_type_id: '$value.id'
+            }
+          };
+
+          if (caseTypeIds) {
+            if (caseTypeIds.length === 0) {
+              return $q.resolve([]);
+            }
+            params.id = { IN: caseTypeIds };
+          }
+
+          return civicaseCrmApi('CaseType', 'get', params).then(function (data) {
+            return getFormattedAwardDetailsData(data.values);
+          });
+        });
+    }
+
+    /**
+     * @param {object} workflow workflow object
+     * @returns {Promise} promise
+     */
+    function createDuplicate (workflow) {
+      return civicaseCrmApi('CaseType', 'create', _.extend({}, workflow, { id: null }))
+        .then(function (data) {
+          var awardsDetailsData = _.clone(workflow.awardDetails);
+
+          awardsDetailsData.id = null;
+          awardsDetailsData.case_type_id = data.values[0].id;
+
+          return civicaseCrmApi([
+            ['AwardDetail', 'create', awardsDetailsData]
+          ]);
+        });
+    }
+
+    /**
+     * @param {object[]} workflows list of workflows
+     * @returns {object[]} formatted list of workflows
+     */
+    function getFormattedAwardDetailsData (workflows) {
+      var managerContacts = [];
+
+      workflows = _.map(workflows, function (workflow) {
+        workflow.awardDetails = workflow['api.AwardDetail.get'].values[0];
+        delete workflow['api.AwardDetail.get'];
+
+        managerContacts = managerContacts.concat(workflow.awardDetails.award_manager);
+
+        workflow.awardDetailsFormatted = {
+          subtypeLabel: awardSubtypes[workflow.awardDetails.award_subtype].label,
+          managers: []
+        };
+
+        return workflow;
+      });
+
+      return ContactsCache.add(managerContacts)
+        .then(function () {
+          return _.map(workflows, function (workflow) {
+            _.each(workflow.awardDetails.award_manager, function (managerID) {
+              workflow.awardDetailsFormatted.managers.push(
+                ContactsCache.getCachedContact(managerID).display_name
+              );
+            });
+
+            return workflow;
```

Comment:
this is returning an array of promises, not an array of objects. I'm not even sure how this is working without breaking.

Probably some AngularJS magic working in the background that waits for all the promises to be completed. In normal javascript you would have to use `Promise.all` for this to work.

Reply:
No wait, this is a array of promises. This is a promise(from Contacts.add), which resolves to array of workflows.